tracer/src/Datadog.Trace/Debugger/Expressions/ProbeProcessor.cs

```diff
@@ -266,36 +273,87 @@ private ExpressionEvaluationResult Evaluate(DebuggerSnapshotCreator snapshotCrea
                 return evaluationResult;
             }
 
+            CheckSpanDecoration(snapshotCreator, ref shouldStopCapture, evaluationResult);
+
             if (evaluationResult.HasError)
             {
                 return evaluationResult;
             }
 
-            if (evaluationResult.Condition is false)
+            if (evaluationResult.Metric.HasValue)
             {
-                // if the expression evaluated to false, stop capture
+                LiveDebugger.Instance.SendMetrics(ProbeInfo.MetricKind.Value, ProbeInfo.MetricName, evaluationResult.Metric.Value);
+                // snapshot creator is created for all probes in the method invokers,
+                // if it is a metric probe, once we sent the value, we can stop the invokers and dispose the snapshot creator
+                snapshotCreator.Dispose();
                 shouldStopCapture = true;
-                return evaluationResult;
             }
 
-            if (evaluationResult.Condition is true &&
-                !ProbeRateLimiter.Instance.Sample(ProbeInfo.ProbeId))
+            if (evaluationResult.Condition != null && // meaning not metric, span probe or span decoration
+                (evaluationResult.Condition is false ||
+                !ProbeRateLimiter.Instance.Sample(ProbeInfo.ProbeId)))
             {
-                // if condition is true, stop capture if there is a rate limit
+                // if the expression evaluated to false, or there is a rate limit, stop capture
                 shouldStopCapture = true;
                 return evaluationResult;
             }
 
-            if (evaluationResult.Metric.HasValue)
+            return evaluationResult;
+        }
+
+        private void CheckSpanDecoration(DebuggerSnapshotCreator snapshotCreator, ref bool shouldStopCapture, ExpressionEvaluationResult evaluationResult)
+        {
+            if (evaluationResult.Decorations == null)
+            {
+                return;
+            }
+
+            const string dynamicPrefix = "_dd.di.";
```

Comment:
It feels a bit hidden, what do you think about lifting it to the top of the file?

Reply:
Fixed in c84d482233d0cffc75af6ff48f08583e125fdcc0